Add support for listing and getting repository/organization webhook deliveries

mumoshu opened the pull request:
Resolves #1933

To make this PR minimal as possible for ease of reviewing, I intentionally included only four APIs related to the webhook deliveries API:

- Listing repository hook deliveries
- Getting a specific repository hook delivery
- Listing organization hook deliveries
- GEtting a specific organization hook delivery

That being said, the following methods aren't covered in this PR. If you like, I'd gladly submit subsequent PRs to add all of them.

- Trigger redelivery of a hook delivery
- ~~List/get/redelivery for organizational webhooks~~

Note that changes on the following files are completely generated by running `go generate` as documented in your guide, so you won't need to read them that carefully.

- `github/github-accessors_test.go`
- `github/github-accessors.go`

FWIW, you can see how these additional methods can be used to retrieve historical webhook deliveries here:

https://github.com/actions-runner-controller/actions-runner-controller/blob/c3cfab81c85d33c65f7b6f6443da0d2f57fd72db/pkg/githubwebhookdeliveryforwarder/githubwebhookdelivery.go#L96-L139

The part that propagates `Cursor` from the response to the ListCursorOptions for the next paginated request is beneficial to understanding how the cursor can be used for pagination.

mumoshu commented on github/github.go:
```diff
@@ -212,6 +212,9 @@ type ListCursorOptions struct {
 	Before string `url:"before,omitempty"`
+
+	// A cursor, as given in the Link header. If specified, the query continues the search using this cursor.
+	Cursor string `url:"cursor,omitempty"`
```
I believe this is the first GitHub API that uses `Cursor` for pagination. And I've added it to `ListCursorOptions` rather than `ListOptions` as... the name seemed to state that it wants all the possible list options for GitHub APIs that support pagination.

Please tell me if you'd like a brand new struct like `ListHookDeliveriesOptions`. Honestly, I'd personally prefer that as the Webhook Deliveries API as of today doesn't support other fields defined in ListCursorOptions. But go-github's current design looked more inclined to have a fewer number of common structs, rather than having a lot of specific structs. That's why I've chosen to change ListCursorOptions rather than having a new struct.

gmlewis replied:
I think adding it to `ListCursorOptions` is fine.  Thanks.

mumoshu commented on github/repos_hooks_deliveries.go:
```diff
@@ -0,0 +1,205 @@
+	return h, resp, nil
+}
+
+func (d *HookDelivery) ParseRequestPayload() (interface{}, error) {
```
This is a utility function to ease parsing event payloads, and it is derived from the equivalent function for Events API https://github.com/google/go-github/blob/05e95d3c51abdf832f9430209ccb46b494ac7be8/github/event.go#L29-L31

gmlewis replied:
Why can't you reuse https://github.com/google/go-github/blob/master/github/event.go#L29-L132 here?

mumoshu replied:
@gmlewis Hey! Interesting idea. How would you like to do that? That snippet in event.go switch on `Type`, where each value can be e.g. `CheckRunEvent`. On the other hand, this one switches on `event` field, where each value is e.g. `check_run`. They're very similar but different.

mumoshu replied:
Also note that this is almost fully covered by `TestHookDelivery_ParsePayload`, if that helps.

gmlewis replied:
What I'm thinking here is to make a helper function that both `ParsePayload` and `ParseRequestPayload` both call.

The helper method could take an anonymous function to perform the required mapping... one of them would use the `eventTypeMethod` map found in [messages.go](https://github.com/google/go-github/blob/master/github/messages.go#L45).

Let me know if you don't think this is possible and I'll download your PR and see if I can get it to work.

gmlewis replied:
Here's my proposed replacement for this method:

```go
// ParseRequestPayload parses the request payload. For recognized event types,
// a value of the corresponding struct type will be returned.
func (d *HookDelivery) ParseRequestPayload() (interface{}, error) {
	eType, ok := eventTypeMapping[*d.Event]
	if !ok {
		return nil, fmt.Errorf("unsupported event type %q", *d.Event)
	}

	e := &Event{Type: &eType, RawPayload: d.Request.RawPayload}
	return e.ParsePayload()
}
```

mumoshu commented on github/repos_hooks_deliveries.go:
```diff
@@ -0,0 +1,205 @@
+// ListHookDeliveries lists webhook deliveries for a webhook configured in a repository.
+//
+// GitHub API docs: https://docs.github.com/en/rest/reference/repos#list-deliveries-for-a-repository-webhook
+func (s *RepositoriesService) ListHookDeliveries(ctx context.Context, owner, repo string, id int64, opts *ListCursorOptions) ([]*HookDelivery, *Response, error) {
```
I"ve chosen to use the term `Hook` rather than `Webhook` across all the changes following  the fact that you seems to prefer the former according to the fact that there are `GetHook` for [Get repository webhooks API](https://docs.github.com/en/rest/reference/repos#get-a-repository-webhook) and `ListHooks` for [List repository webhooks API](https://docs.github.com/en/rest/reference/repos#list-repository-webhooks) and so on.

If you prefer the term `Webhook` or another one for some reason, please let me know!

mumoshu commented on github/repos_hooks_deliveries.go:
```diff
@@ -0,0 +1,205 @@
+
+type HookRequest struct {
+	Header     map[string]string `json:"headers,omitempty"`
+	RawPayload *json.RawMessage  `json:"payload,omitempty"`
```
I intentionally added `Raw-` prefix to the field name following the fact that the very similar `payload` field in the event object uses `RawPayload` https://github.com/google/go-github/blob/7a077bbd7313615d4f924f12b065913ee94e1d58/github/event.go#L13-L23

gmlewis commented on github/github.go:
```diff
@@ -481,25 +489,33 @@ func (r *Response) populatePageValues() {
 			}
-			page := url.Query().Get("page")
-			if page == "" {
+
```
Hmmm... I'm not sure I'm happy with this section (especially regarding increasing the already-deep indentation levels).

Since `"cursor"` is the less-common and smaller case, let's please handle it first, and let's revert all the changes made to the handling of the `"page"` section, except maybe to pull out the shared `query` first... something like this:

```go
q := url.Query()
if cursor := q.Get("cursor"); cursor != "" {
...
}

page := q.Get("page")
if page == "" {
  continue
}
...
```

mumoshu replied:
Thanks! Addressed in f6d688a.

gmlewis commented on github/repos_hooks_deliveries.go:
```diff
@@ -0,0 +1,205 @@
+type HookDelivery struct {
+	ID             *int64     `json:"id"`
+	GUID           *string    `json:"guid"`
+	DeliveredAt    *time.Time `json:"delivered_at"`
```
```suggestion
	DeliveredAt    *Timestamp `json:"delivered_at"`
```

mumoshu replied:
Addressed in https://github.com/google/go-github/pull/1934/commits/ff5a47a9e3d968697afa6cf425a5220f774f9c9c !

gmlewis commented on github/repos_hooks_deliveries.go:
```diff
@@ -0,0 +1,205 @@
+	GUID           *string    `json:"guid"`
+	DeliveredAt    *time.Time `json:"delivered_at"`
+	Redelivery     *bool      `json:"redelivery"`
+	Duration       *float64   `json:"duration"`
```
Shouldn't this be `*int64` ?  Where is this documented?
Neither of the URLs above have a `duration` field on their pages.

mumoshu replied:
It isn't documented but looks like the below example. I've grabbed this by calling the list deliveries API on my own repo:

```
  {
    "id": 12076761161,
    "guid": "5a73eedc-d706-11eb-9a56-eb84e3dea363",
    "delivered_at": "2021-06-27T05:13:05Z",
    "redelivery": false,
    "duration": 0.29,
    "status": "Invalid HTTP Response: 530",
    "status_code": 530,
    "event": "push",
    "action": null,
    "installation_id": null,
    "repository_id": 366236975
  },
```

gmlewis replied:
Awesome. Thank you. Please ignore this comment of mine.

gmlewis commented on github/repos_hooks_deliveries.go:
```diff
@@ -0,0 +1,205 @@
+}
+
+type HookResponse struct {
+	Header     map[string]string `json:"headers,omitempty"`
```
```suggestion
	Headers     map[string]string `json:"headers,omitempty"`
```

mumoshu replied:
Fixed in https://github.com/google/go-github/pull/1934/commits/dffbb1cd1b1975ba93f77fb2d7331dffae56d198

gmlewis commented on github/repos_hooks_deliveries.go:
```diff
@@ -0,0 +1,205 @@
+//
+// GitHub API docs: https://docs.github.com/en/rest/reference/repos#list-deliveries-for-a-repository-webhook
+func (s *RepositoriesService) ListHookDeliveries(ctx context.Context, owner, repo string, id int64, opts *ListCursorOptions) ([]*HookDelivery, *Response, error) {
+	u := fmt.Sprintf("repos/%v/%v/hooks/%d/deliveries", owner, repo, id)
```
```suggestion
	u := fmt.Sprintf("repos/%v/%v/hooks/%v/deliveries", owner, repo, id)
```

mumoshu replied:
Fine! But out of curiosity, why do you prefer using `%v` for an integer value?

mumoshu replied:
Fixed in 5f0e1e4 anyway

gmlewis replied:
I prefer `%v` for the following reasons:

* more consistent with the rest of the repo,
* idiomatic Go,
* and most importantly, it makes it much easier to search for an existing endpoint when you don't have to figure out the types of the segments of the URLs... you can more easily transform a URL from the documentation into a search string into the code base.

mumoshu replied:
Gotcha! That totally makes sense. Thanks a lot for clarifying.

Regarding consistency with the rest of the repo, JFYI, I found a bunch of `%d` occurrences in the existing sources like `repos_hook.go`. Probably you'd want to fix it in another circumstance.

gmlewis reviewed (changes requested):
Thank you, @mumoshu.
I'm thinking that this can be improved by reusing more of the existing event code or refactoring it so that it is easier to reuse.

Also, the drop in code coverage is a bit concerning, but this might get fixed by reducing the amount of new code and reusing more of the existing code.

ohpintu90 commented:
Ok

mumoshu commented:
@gmlewis Hey! I've addressed all the comments. Would you mind reviewing once again?

Also, please feel free to squash on merge, or just tell me to do so, if necessary. Thanks!

mumoshu commented:
FYI: I've added 431c77c for improving test coverage

mumoshu commented:
I've added 23922a8 for more code coverage on `cursor` handling.

All the new codes should have been covered. I have no immediate plan to add more commits now so it should be a good timing for you to review again 😄

mumoshu commented:
Apparently, the last commit 23922a8 seems not included in https://app.codecov.io/gh/google/go-github/branch/webhook-deliveries-api yet. But theoretically speaking the code coverage shouldn't get decreased now, as I've already added enough tests to cover all the new code paths.

gmlewis commented:
> Also, please feel free to squash on merge, or just tell me to do so, if necessary. Thanks!

We always squash and merge in this repo, so don't worry about the number of commits.  In fact, we prefer that author's don't force-push so that it is easier for reviewers to see what changes have been made since the last review.

gmlewis commented on github/github.go:
```diff
@@ -481,7 +489,19 @@ func (r *Response) populatePageValues() {
+						r.Cursor = cursor
+					}
+				}
+			}
```
Are `"cursor"` and `"page"` mutually exclusive?  (I think the answer is "yes" but I could be wrong.)

If so, do you want to add a `continue` within this `if` block just to make that clear?
I'm actually fine either way... your call.

mumoshu replied:
Thanks! Yes, I believe so, although it isn't documented anywhere. Added `continue` in 72c3380

gmlewis commented on github/repos_hooks_deliveries.go:
```diff
@@ -0,0 +1,204 @@
+	InstallationID *string    `json:"installation_id"`
+	RepositoryID   *int64     `json:"repository_id"`
+
+	// Request is populated by GetHookDelivery
```
nit:

```suggestion
	// Request is populated by GetHookDelivery.
```

mumoshu replied:
Addressed in 5b86aff 👍

gmlewis commented on github/repos_hooks_deliveries.go:
```diff
@@ -0,0 +1,204 @@
+
+	// Request is populated by GetHookDelivery
+	Request *HookRequest `json:"request,omitempty"`
+	// Response is populated by GetHookDelivery
```
nit:

```suggestion
	// Response is populated by GetHookDelivery.
```

mumoshu replied:
Addressed in 5b86aff 👍

gmlewis commented on github/repos_hooks_deliveries.go:
```diff
@@ -0,0 +1,204 @@
+	return Stringify(d)
+}
+
+type HookRequest struct {
```
Can you please add a comment here describing `// HookRequest ...` ?

mumoshu replied:
Added in 194eff6 👍

gmlewis commented on github/repos_hooks_deliveries.go:
```diff
@@ -0,0 +1,204 @@
+	return Stringify(r)
+}
+
+type HookResponse struct {
```
Can you please add a comment here describing `// HookResponse ...` ?

mumoshu replied:
Added in 194eff6 👍

mumoshu commented:
> What I'm thinking here is to make a helper function that both ParsePayload and ParseRequestPayload both call.

Sounds good!

> The helper method could take an anonymous function to perform the required mapping... one of them would use the eventTypeMethod map found in messages.go.

And would you mind if I also created another, reverse map of `eventTypeMethod`, named like `typeEventMapping`, and use that from `ParsePayload`?

mumoshu commented:
Apparently, another method named `ParseWebhook` that calls `ParsePayload` seems to already use `eventTypeMapping` on its own 🤔 Probably adding second use of `eventTypeMapping` in an anonymous function defined in `ParseWebhook` doesn't make much sense?

Perhaps we'd better revise `ParseWebHook` as well? I'm pretty lost on how I should go.

gmlewis commented:
> I'm pretty lost on how I should go.

OK. I'll take a look and get back to you.

gmlewis commented:
> OK. I'll take a look and get back to you.

I think I've got a simple solution for you.  It also caught 3 errors in your test table, which I think justifies the reuse of the existing code.  I'll add review comments so you can see what I did.

Let me know what you think.

gmlewis commented on github/repos_hooks_deliveries_test.go:
```diff
@@ -0,0 +1,220 @@
+	"content_reference":              &ContentReferenceEvent{},
+	"create":                         &CreateEvent{},
+	"delete":                         &DeleteEvent{},
+	"deploy_ket":                     &DeployKeyEvent{},
```
```suggestion
	"deploy_key":                     &DeployKeyEvent{},
```

gmlewis commented on github/repos_hooks_deliveries_test.go:
```diff
@@ -0,0 +1,220 @@
+	"issues":                         &IssuesEvent{},
+	"label":                          &LabelEvent{},
+	"marketplace_purchase":           &MarketplacePurchaseEvent{},
+	"member_event":                   &MemberEvent{},
```
```suggestion
	"member":                   &MemberEvent{},
```

gmlewis commented on github/repos_hooks_deliveries_test.go:
```diff
@@ -0,0 +1,220 @@
+	"label":                          &LabelEvent{},
+	"marketplace_purchase":           &MarketplacePurchaseEvent{},
+	"member_event":                   &MemberEvent{},
+	"membership_event":               &MembershipEvent{},
```
```suggestion
	"membership":               &MembershipEvent{},
```

gmlewis commented on github/repos_hooks_deliveries_test.go:
```diff
@@ -0,0 +1,220 @@
+	}
+
+	_, err := d.ParseRequestPayload()
+	if err == nil || err.Error() != "unexpected end of JSON input" {
```
```suggestion
	if err == nil || err.Error() != `unsupported event type "some_invalid_event"` {
```

gmlewis reviewed (changes requested):
Hi @mumoshu - here are my suggestions for reusing existing code, along with a few error corrections. Please let me know if they make sense to you.

Please remember to run `gofmt` after accepting my suggestions, as they will need proper formating.

mumoshu commented:
@gmlewis Thank you so much for taking your time and the suggestions! All looked awesome and I've included your suggestions in the last commit, 1381c0f.

Regarding the change on `ParseRequestPayload`, I initially had some preconceived idea that it won't be acceptable to create a new `Event` object just for calling ParsePayload. Apparently, it was the best impl though. Good job!

gmlewis reviewed (approved):
Thank you, @mumoshu !
LGTM.

Awaiting second LGTM before merging.

mumoshu commented:
@gmlewis Hey! Would you mind if I added another commit for adding similar support for organization variants of these APIs?

https://github.com/mumoshu/go-github/commit/0a3739c94ba1eb5fe9f040f465b1e5387eb802e1.

It already has full test coverage, and I've manually verified it to work as part of https://github.com/actions-runner-controller/actions-runner-controller/pull/682, too.

mumoshu commented:
@gmlewis Hey! Would you mind if I added another commit for adding similar support for organization variants of these APIs?

https://github.com/mumoshu/go-github/commit/0a3739c94ba1eb5fe9f040f465b1e5387eb802e1.

It already has full test coverage, and I've manually verified it to work as part of https://github.com/actions-runner-controller/actions-runner-controller/pull/682, too.

@Parker77 Hi! If you're reading this and going to put the second LGTM, please go ahead!

I'm just trying to reduce the time required on both ends to make the set of changes related to #1933. If you merged this, I'll just submit another PR for https://github.com/mumoshu/go-github/commit/0a3739c94ba1eb5fe9f040f465b1e5387eb802e1.

gmlewis commented:
> @gmlewis Hey! Would you mind if I added another commit for adding similar support for organization variants of these APIs?

Hi @mumoshu - sure, you can add more commits if you wish... please just push and don't use "force push" so that it will be easier to review more recent additions. Thank you.

mumoshu commented:
@gmlewis Thanks for confirming! I've added 0a3739c

gmlewis commented on github/orgs_hooks_deliveries.go:
```diff
@@ -0,0 +1,48 @@
+package github
```
Please add the copyright header to this new file (with year 2021 and a blank line before the `package` line).

mumoshu replied:
Addressed in fc8d20ffc 👍

gmlewis commented on github/orgs_hooks_deliveries.go:
```diff
@@ -0,0 +1,48 @@
+	if err != nil {
+		return nil, nil, err
+	}
+	h := new(HookDelivery)
```
Please add a blank line above this line for consistency with this repo and idiomatic Go (after an `if err` block).

mumoshu replied:
Thanks! Addressed in 123bafa

gmlewis reviewed (changes requested):
Thank you, @mumoshu !
Just a couple minor tweaks, please.

gmlewis reviewed (approved):
Thank you, @mumoshu !
LGTM.

Awaiting second LGTM before merging.

Parker77 reviewed (approved):
LGTM.
